File: perf/benchmarks/apps/spectral_norm/spectral_norm.go
// Spectral norm — CLBG numerical kernel (Go).
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

func a(i, j int) float64 {
	return 1.0 / float64(((i+j)*(i+j+1)>>1)+i+1)
}

func av(v, out []float64, n int) {
	for i := 0; i < n; i++ {
		s := 0.0
		for j := 0; j < n; j++ {
			s += a(i, j) * v[j]
		}
		out[i] = s
	}
}

func atv(v, out []float64, n int) {
	for i := 0; i < n; i++ {
		s := 0.0
		for j := 0; j < n; j++ {
			s += a(j, i) * v[j]
		}
		out[i] = s
	}
}

func atav(v, out, tmp []float64, n int) {
	av(v, tmp, n)
	atv(tmp, out, n)
}

func main() {
	n := 1500
	if len(os.Args) > 1 {
		v, _ := strconv.Atoi(os.Args[1])
		n = v
	}
	u := make([]float64, n)
	v := make([]float64, n)
	tmp := make([]float64, n)
	for i := 0; i < n; i++ {
		u[i] = 1.0
	}
	t0 := time.Now()
	for i := 0; i < 10; i++ {
		atav(u, v, tmp, n)
		atav(v, u, tmp, n)
	}
	vbv, vv := 0.0, 0.0
	for i := 0; i < n; i++ {
		vbv += u[i] * v[i]
		vv += v[i] * v[i]
	}
	sn := math.Sqrt(vbv / vv)
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	cs := int64(math.Round(sn * 1e9))
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/spectral_norm/spectral_norm.rs
// Spectral norm — CLBG numerical kernel (Rust).
use std::env;
use std::time::Instant;

#[inline]
fn a(i: i64, j: i64) -> f64 {
    1.0 / ((((i + j) * (i + j + 1)) >> 1) + i + 1) as f64
}

fn av(v: &[f64], out: &mut [f64], n: usize) {
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n {
            s += a(i as i64, j as i64) * v[j];
        }
        out[i] = s;
    }
}

fn atv(v: &[f64], out: &mut [f64], n: usize) {
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n {
            s += a(j as i64, i as i64) * v[j];
        }
        out[i] = s;
    }
}

fn atav(v: &[f64], out: &mut [f64], tmp: &mut [f64], n: usize) {
    av(v, tmp, n);
    atv(tmp, out, n);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 { args[1].parse().unwrap_or(1500) } else { 1500 };
    let mut u: Vec<f64> = vec![1.0; n];
    let mut v: Vec<f64> = vec![0.0; n];
    let mut tmp: Vec<f64> = vec![0.0; n];

    let t0 = Instant::now();
    for _ in 0..10 {
        atav(&u, &mut v, &mut tmp, n);
        atav(&v, &mut u, &mut tmp, n);
    }
    let mut vbv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        vbv += u[i] * v[i];
        vv += v[i] * v[i];
    }
    let sn = (vbv / vv).sqrt();
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    let cs = (sn * 1e9).round() as i64;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
